### DISCUSSION TITLE
Simple simulation of electrical potential / current.

### DISCUSSION THREAD

**Discussion by Drinausaur** [mfem/mfem#3505] [Q&A]
Hello,

I am not a FEM expert, but an electronics guy who reads about FEM online ( including the MFEM guide ).

I am trying to set up a "simple" simulation of some pieces of copper, add voltage sources ( DBC if I understand properly ), and current sources ( NBC ).

I started with the voltage sources starting from example 1, but I can't figure out why it does not work. ( I have been looking at that code for days now ).
It seems that I am able to set the voltage to some regions, but not all, and the regions that are assigned take different values.

Before that code, I load the mesh, store voltage values in `m_dbcv` and the region they correspond to in `m_dbcs`

```
    new( &m_fec ) mfem::H1_FECollection( 1, m_mesh.Dimension() );

    new( &m_fespace ) mfem::FiniteElementSpace( &m_mesh, &m_fec );
    Report( "Number of unknowns: " + std::to_string( m_fespace.GetTrueVSize() ) );

    // Set conductivity
    mfem::ConstantCoefficient conductivity( 1 / 1.68e-8 ); // 1 / ( ohm.m )
    mfem::ConstantCoefficient one(1.0);
    mfem::ConstantCoefficient zero(0);

    // Set the Dirichlet values in the solution vector
    mfem::GridFunction u(& m_fespace );
    u = 0.0;

    Report( "Number of element attributes: " + std::to_string(m_mesh.attributes.Size()) );

    // Prepare the source term in the right-hand-side
    mfem::LinearForm b(&m_fespace);
    b.AddDomainIntegrator(new mfem::DomainLFIntegrator( one ));

    // Define the Dirichlet Boundary conditions
    mfem::Array<int> ess_tdof_list;

    for (int i=0; i< m_dbcs.Size(); i++)
    {
        Report( "Applying Dirichlet on region " + std::to_string( m_dbcs[i] ) + " : " + std::to_string( m_dbcs[i] )  );
        mfem::Array<int> dirichlet_attribute =
        dirichlet_attribute.Append( m_dbcs[i] );
        int dirichlet_value = m_dbcv[i];
        mfem::Array<int> dirichlet_tdof_list;
        m_fespace.GetEssentialTrueDofs( dirichlet_attribute, dirichlet_tdof_list);
        ess_tdof_list.Append( dirichlet_tdof_list );
        mfem::ConstantCoefficient bc_coeff(dirichlet_value);
        u.ProjectBdrCoefficient(bc_coeff, dirichlet_tdof_list);
    }

    b.Assemble();

    // Prepare the bilinear form
    mfem::BilinearForm a(&m_fespace);
    a.AddDomainIntegrator(new mfem::DiffusionIntegrator( conductivity ));
    a.Assemble();

    // Prepare the linear system with enforcement of the essential boundary
    // conditions
    mfem::SparseMatrix A;
    mfem::Vector B, X;
    a.FormLinearSystem(ess_tdof_list, u, b, A, X, B);

    // 9. Solve the system using PCG with symmetric Gauss-Seidel preconditioner.
    mfem::GSSmoother M(A);
    mfem::PCG(A, M, B, X, 1, 200, 1e-12, 0.0);

    // Recover the solution
    a.RecoverFEMSolution(X, b, u );
    u.Save("sol.gf");
```
And then I get this: ( I tried to set each element to a different voltage, I did not connected them, yet )

![image](https://user-images.githubusercontent.com/45109319/220475006-ff606117-23b2-414c-8e86-165487de0426.png)

here is my mesh:
[pcb_3d.msh.zip](https://github.com/mfem/mfem/files/10798375/pcb_3d.msh.zip)

Could you help me understand why it does not work?

**Comment by mlstowell** [comment]
Hello, @Drinausaur,

Unfortunately, this is a common pitfall when working with attributes. We need to document this better. The array you pass to `GetEssentialTrueDofs()` is not meant to be the set of attributes of interest. Instead it is a "marker" array. Its length should be `mesh.attributes.Max()` and it should contain zeros and ones where the ones indicate the attributes of interest by their position in the array.

That being said, there is another issue: `GetEssentialTrueDofs()` works with "element" attributes as opposed to "boundary" attributes. I'm not sure which you intended. Also, `ProjectBdrCoefficient` does not expect true DoF indices but rather a "boundary attribute marker array". This is a "marker" array as described above but its length is `mesh.bdr_attributes.Max()`.

I would suggest avoiding `GetEssentialTrueDofs()` because it probably isn't needed in this context. Start by creating a marker array using either:
```
mfem::Array<int> dirichlet_marker(m_mesh.attributes.Max());
dirichlet_marker = 0; dirichlet_marker[m_dbcs[i]-1] = 1;
u.ProjectCoefficient(bc_coeff, dirichlet_marker);
```
or
```
mfem::Array<int> dirichlet_bdr_marker(m_mesh.bdr_attributes.Max());
dirichlet_bdr_marker = 0; dirichlet_bdr_marker[m_dbcs[i]-1] = 1;
u.ProjectBdrCoefficient(bc_coeff, dirichlet_bdr_marker);
```
With one of these inside your loop you should set the proper values in `u`.

Best wishes,
Mark

**Comment by Drinausaur** [comment]
Thank you for your answer,

I made some changes to my code with your help, and it certainly eliminated some mistakes, but the simulation is not yet as expected. I suppose I messed up with the equations.



```
    new( &m_fec ) mfem::H1_FECollection( 1, m_mesh.Dimension() );

    new( &m_fespace ) mfem::FiniteElementSpace( &m_mesh, &m_fec );
    Report( "Number of unknowns: " + std::to_string( m_fespace.GetTrueVSize() ) );

    // Set conductivity
    mfem::ConstantCoefficient conductivity( 1 / 1.68e-8 ); // 1 / ( ohm.m )
    mfem::ConstantCoefficient one(1.0);
    mfem::ConstantCoefficient zero(0);
    mfem::GridFunction phi(& m_fespace );
    
    Report( "Running solver... " );
    // Initialize the electric potential with its boundary conditions
    phi = 0.0;
    

    // Prepare the source term in the right-hand-side
    mfem::LinearForm b(&m_fespace);
    b.AddDomainIntegrator(new mfem::DomainLFIntegrator( one ));

    mfem::Array<int> dbc_bdr_attr(m_mesh.bdr_attributes.Max());
    for ( int i=0; i < m_dbcs.Size(); i++ )
    {
        Report( "Trying to set voltage on region " + std::to_string( m_dbcs[i] ) + " : " + std::to_string( m_dbcv[i] ) );
        mfem::ConstantCoefficient voltage(m_dbcv[i]);
        mfem::Array<int> dirichlet_bdr_marker(m_mesh.bdr_attributes.Max());
        dirichlet_bdr_marker = 0; dirichlet_bdr_marker[m_dbcs[i]-1] = 1;
        phi.ProjectBdrCoefficient(voltage, dirichlet_bdr_marker);
    }

    b.Assemble();

    // Prepare the bilinear form
    mfem::BilinearForm a(&m_fespace);
    a.AddDomainIntegrator(new mfem::DiffusionIntegrator( conductivity ));
    a.Assemble();

    mfem::Array<int> ess_tdof_list(0);
    mfem::Array<int> ess_bdr(m_mesh.bdr_attributes.Max());
    ess_bdr = 0;

    // Prepare the linear system with enforcement of the essential boundary
    // conditions
    mfem::OperatorPtr A;
    mfem::Vector B, X;
    a.FormLinearSystem(ess_tdof_list, phi, b, A, X, B);

    // Use a simple symmetric Gauss-Seidel preconditioner with PCG.
    mfem::GSSmoother M((mfem::SparseMatrix&)(*A));
    mfem::PCG(*A, M, B, X, 1, 200, 1e-12, 0.0);
      
    // Recover the solution
    a.RecoverFEMSolution(X, b, phi );
    phi.Save("sol.gf");
```

Just for some context, I am working on an open source PCB editor software ( KiCad ) and we'd like to add some FEM capabilities.

For now, the thing we'd like to run is a simple voltage / current simulation, where we can set voltage sources ( so dirichlet BC ) and current sources ( Neumann BC I suppose ) and then let the resistivity  parameter determine the gradient electric potential everywhere.

So in that example here, what I'm trying to achieve is setting some "pillars" to a potential ( so they should be uniformly colored ), and the current that results should create a potential gradient along connections.
![image](https://user-images.githubusercontent.com/45109319/223370006-795756db-c98f-4821-9beb-8a71b18f701a.png)

We have made some successful test with another solver, but we'd like to test mfem.
![image](https://user-images.githubusercontent.com/45109319/223363230-499e3a1f-c354-4ceb-9dd1-27c41f4a1939.png)


And this is the equation we'd like to solve: ( this is from my notes when working with the other solver )
At the moment, we focus on DC conduction as this is the certainly the simplest FEM simulation we can run.
![image](https://user-images.githubusercontent.com/45109319/223364822-5e3d316f-00ff-4401-a939-3e0f653c7b6f.png)

**Reply by mlstowell** [reply]
It looks like the correct Dirichlet values should be making their way into `phi`. The next step is to adjust the system matrix and right-hand-side vector so that the solver will use these values correctly. This is done in the call to `FormLinearSystem`. For this to work the array in the first argument, `ess_tdof_list` in this case, needs to contain the degree of freedom indices for all Dirichlet values. This is usually done with a call to `GetEssentialTrueDofs`, see `ex1.cpp`. It looks like you have two possible candidates for collecting the necessary information; `dbc_bdr_attr` and `ess_bdr`. I would suggest setting `dbc_bdr_attr` to zero before the loop  and then setting `dbc_bdr_attr[m_dbcs[i]-1]= 1` within the loop. Then after the loop you should have marked all of the Dirichlet boundaries. Then someplace after the loop but before `FormLinearSystem` call `m_fespace.GetEssentialTrueDofs(dbc_bdr_attr, ess_tdof_list)`. That should solve the system with the desired Dirchlet values applied.

One other thing I notice is that your right-hand-side is being set to 1. This implies a uniform charge density equal to 1 everywhere in the domain which is probably not what you want. I would remove the call to `AddDomainIntegrator` and just set `b = 0;`. Later, when you're ready to add Neumann BCs you can call `AddBoundaryIntegrator` with the appropriate arguments, `ex27.cpp` might be helpful when you get to that point.

**Reply by Drinausaur** [reply]
Thank you so much! The simulation is starting to go somewhere. The last thing I can notice, for now, is that the "pillars" I don't set seem to get a zero voltage value. In my simulation, there should not be any current (hence no gradient) in the branches with unset extremities.

![image](https://user-images.githubusercontent.com/45109319/223870264-0af5a638-f3db-47f5-869f-c3d6fd2338e8.png)

(btw my log looks like this, it does not set those regions to 0) :
```
Setting Dirichlet BC on region 2: 2
Setting Dirichlet BC on region 3: 3
Setting Dirichlet BC on region 4: 1
Running solver... 
Setting voltage on region 2 : 2.000000
Setting voltage on region 3 : 3.000000
Setting voltage on region 4 : 1.000000

```
I'm leaving the code here, so it can maybe help someone in the future:
( I tried changing the `phi=0.0` line to something else like `phi=2.5` but with no success )

```

    // 3. Define a finite element space on the mesh. Here we use H1 continuous
    //    high-order Lagrange finite elements of the given order.
    new( &m_fec ) mfem::H1_FECollection( 1, m_mesh.Dimension() );
    new( &m_fespace ) mfem::FiniteElementSpace( &m_mesh, &m_fec );

    // Set conductivity
    mfem::ConstantCoefficient conductivity( 1.0 / 1.68e-8 ); // 1 / ( ohm.m )
    mfem::GridFunction phi(& m_fespace );
    
    Report( "Running solver... " );
    // Initialize the electric potential with its boundary conditions
    phi = 0.0;
    

    // Prepare the source term in the right-hand-side
    mfem::LinearForm b(&m_fespace);
    b = 0.0;
    b.Assemble();

    mfem::Array<int> dbc_bdr_attr(m_mesh.bdr_attributes.Max());
    dbc_bdr_attr = 0;

    for ( int i=0; i < m_dbcs.Size(); i++ )
    {
        Report( "Setting voltage on region " + std::to_string( m_dbcs[i] ) + " : " + std::to_string( m_dbcv[i] ) );
        mfem::ConstantCoefficient voltage(m_dbcv[i]);
        mfem::Array<int> dirichlet_bdr_marker(m_mesh.bdr_attributes.Max());
        dirichlet_bdr_marker = 0;
        dirichlet_bdr_marker[m_dbcs[i]-1] = 1;
        dbc_bdr_attr[m_dbcs[i]-1]= 1;

        phi.ProjectBdrCoefficient(voltage, dirichlet_bdr_marker);
    }

    // Prepare the bilinear form
    mfem::BilinearForm a(&m_fespace);
    a.AddDomainIntegrator(new mfem::DiffusionIntegrator( conductivity ));
    a.Assemble();

    mfem::Array<int> ess_tdof_list;
    m_fespace.GetEssentialTrueDofs(dbc_bdr_attr, ess_tdof_list);

    // Prepare the linear system with enforcement of the essential boundary
    // conditions
    mfem::OperatorPtr A;
    mfem::Vector B, X;
    a.FormLinearSystem(ess_tdof_list, phi, b, A, X, B);

    // Use a simple symmetric Gauss-Seidel preconditioner with PCG.
    mfem::GSSmoother M((mfem::SparseMatrix&)(*A));
    mfem::PCG(*A, M, B, X, 1, 200, 1e-12, 0.0);
      
    // Recover the solution
    a.RecoverFEMSolution(X, b, phi );
    phi.Save("sol.gf");
```

**Reply by jandrej** [reply]
> In my simulation, there should not be any current (hence no gradient) in the branches with unset extremities.

You mean the branch reaching out to a zero pillar should be abrupt zero in the center connection? Since you are using an H1 space there has to be continuity. Which means you will always see a continuous, although rapidly decreasing gradient. To me the field above represents what you described in the code.

**Reply by Drinausaur** [reply]
No, I meant no gradient, the grid function should be continuous. The pillars that are not set should not affect the electric current ( almost no effect) , because I want them to be floating, so no current flowing into/from them. 

If I set a pillar to 0V, and another to 1V, there should be a gradient going from 0V to 1V, the middle should be 0.5V, and all unset pillars should be 0.5V. And for some reason, I could set it to 400V and 401V, having a 0V would change the result in a significant way.

The result I was expecting looks something like that: ( well, in the picture above, I do have 3 DBC, but only 2 in that one )
![image](https://user-images.githubusercontent.com/45109319/223884643-9bb21786-83f0-4735-9778-a363e0f89cd4.png)

**Reply by mlstowell** [reply]
I agree with @Drinausaur. This looks odd to me as well. If I'm interpreting things correctly the solution should vary linearly from pillar 4 to pillar 3. The other six branches and the pillars at their ends should all be at a voltage of 2.

@Drinausaur, can you attach the mesh file and perhaps the complete code? I'm eager to tinker with this myself.

**Reply by mlstowell** [reply]
The earlier image would make sense if the right-hand-side was non-zero...

**Reply by jandrej** [reply]
Oh I see. This looks like an attribute/meshing issue then.

**Reply by Drinausaur** [reply]
Sure, here are the sources ( with the latest commit ):

https://gitlab.com/Drinausaur/kicad/-/tree/ba88e5197cc55336606e374655e6bb3c8c3f5ad9/pcbnew/fem

The mfem part of the code is stored in the `mfem folder` 
The entry point for you would be `bool MFEM_SOLVER::LoadMesh()` ( which actually runs the simulation, despite the name ) in `mfem_solver.cpp`.

And here is the mesh file, it is a `gmsh` file:

[kicad_pcb_3d.msh.zip](https://github.com/mfem/mfem/files/10933801/kicad_pcb_3d.msh.zip)

**Reply by jandrej** [reply]
<img width="1543" alt="Screen Shot 2023-03-09 at 10 23 21 AM" src="https://user-images.githubusercontent.com/5412886/224120016-e951d78b-667c-4d90-97ca-e9b303548c8d.png">

Is this really the mesh you use for the computation? Because I wouldn't expect this to work at all. The connectors have severely stretched triangular elements.

Furthermore, your finite element space is an embedded space. You are computing on 2D elements in 3D. There will probably be quirks to work out in terms of handling boundary conditions etc.

**Reply by Drinausaur** [reply]
Our meshing algorithm is not great and is due for a rework. We usually get some results that are "good enough" by refining the mesh and using actual PCB shapes ( not only straight lines).  But indeed, that one is especially bad.

Here is another one that is using another item and generates a much finer mesh ( but is too fine in some places ), the behaviour is the same ( the set potentials were 5V, 4V and 3V. There should not be a potential higher than 5V or lower than 3V )

![image](https://user-images.githubusercontent.com/45109319/224124607-66f074ac-cfd8-4832-aea7-e73d24c748ef.png)
![image](https://user-images.githubusercontent.com/45109319/224124724-ef5f9460-80a7-425a-91c8-e81bd636cbfa.png)

[kicad_pcb_3d.msh.zip](https://github.com/mfem/mfem/files/10934975/kicad_pcb_3d.msh.zip)

Regarding the embedded space thing, I don't really know what to do with that information. All my elements are in 3D (their thickness is the usual thickness of copper plating, 35µm but the whole square is 10cm x 10cm ) , I added the boundary 2D surfaces for this mfem test in order to apply DBCs, but this is maybe not the way to go

**Reply by jandrej** [reply]
> There should not be a potential higher than 5V or lower than 3V

But you set quite a few Dirichlet boundary attributes to 0V in your code?

**Reply by jandrej** [reply]
You have to make the PCB and the non-powered pillars a free flux (don't apply dirichlet to them). Here you see the initial condition 
<img width="468" alt="Screen Shot 2023-03-09 at 1 49 54 PM" src="https://user-images.githubusercontent.com/5412886/224168058-f371c349-3d12-4faf-a66d-c845ac5d3ed9.png">
and the CONVERGED result. Your CG solver did not converge. I recommend using a proper preconditioner setup (e.g. AMG). This results in 
<img width="512" alt="Screen Shot 2023-03-09 at 1 50 05 PM" src="https://user-images.githubusercontent.com/5412886/224168094-4671d30b-6d1f-4f7a-9615-d8321894c07b.png">
We can help choosing the right setup of course :)

**Reply by Drinausaur** [reply]
Oh I feel dumb not checking for convergence. I can see it in the logs now :/ ( I should add a check in the code )

Is the `AMG` preconditioner `HypreBoomerAMG` ? is there a serial version of that ?

Could you share the lines you changed to get those results ?

is there a resource ( a pdf, a wiki page, etc...) you would recommend that explains how to choose a preconditioner/solver in mfem ? 
Choosing the right setup seems to be one of the hardest part for me, with little FEM knowledge.

**Reply by jandrej** [reply]
AMG in my test is `HypreBoomerAMG`, yes. Unfortunately you'd have to build MFEM with MPI and hypre as dependencies for that. A direct solver (LU, a serial package is UMFPACK for example) would probably work as well, but it will be more memory hungry and only scales "so much" with increasing system sizes.

Choosing a solver is highly problem dependent, there is no definitive answer in general. The resource I'd recommend would also depend on the problem. For a Laplace setup like this, CG+AMG is usually the choice. That being said, if you plan to use "boutique solvers" like AMS/ADS when you want to implement electromagnetic behavior on your PCB models, MFEM is a really good choice :).

**Reply by Drinausaur** [reply]
Could you suggest a working preconditioner/solver for that specific problem that would work with the serial version?

Going with the parallel version in the future is not excluded, but at the moment, I'd need something that works on basic cases.
Moreover, our FEM interface might need some rework to support mfem more easily, might as well do that soon.

The idea is to fix things on our side before adding things that could break it even further such as going parallel ^^'

**Reply by jandrej** [reply]
Direct solvers are probably your best bet and are usually very robust until you reach a certain system size. MFEM is integrated with UMFPACK and MUMPS. Both should be easy to install with a package manager and then you can link them after enabling the option in the MFEM build configuration.

**Reply by Drinausaur** [reply]
sorry, when I wrote solver, I meant those that are integrated in mfem.
The idea is to use mfem in order to limit the number of dependencies. ( therefore using mfem directly )

My question was, after reformulation:

What should I replace those lines with in order to hope for convergence? ( or is there another thing I should replace ? )
```
    // Use a simple symmetric Gauss-Seidel preconditioner with PCG.
    mfem::GSSmoother M((mfem::SparseMatrix&)(*A));
    mfem::PCG(*A, M, B, X, 1, 200, 1e-12, 0.0);
```

**Reply by jandrej** [reply]
Oh I see. You don't want to introduce any TPL. If that is the case your best choice is what you have there. PCG with Gauss-Seidel, which is unfortunately weak for larger Laplace like systems. Increase the number of iterations (here you have `200`) to something very large like 5000 and try to figure out a good initial guess. E.g. when you expect your solution to have a minimum potential of 3.0V then you can set your initial guess in the GridFunction to that.

**Reply by v-dobrev** [reply]
If PCG+Gauss-Seidel starts to get slow as you refine the mesh, you can then try the geometric multigrid solver (assuming you do the mesh refinements in MFEM and not before that in your meshing tool). See `examples/ex26.cpp` for an example of using the geometric multigrid for a diffusion problem.

**Comment by mlstowell** [comment]
I ran your mesh in a modified version of `ex1p` and the results look fine.
![disc_3505](https://user-images.githubusercontent.com/6109571/224170312-3c468584-aa40-4a20-b71a-a166adec459d.png)

While it is an unusual mesh with rather extreme elements I don't think this is the problem. I haven't looked closely at the code yet but perhaps we've missed something there.

**Comment by Drinausaur** [comment]
I have been doing some tests for the past few days, and the simulation is working.
However, I have some current that leaks out of the simulation.
In the picture below, I set a circle to 1V and another one to 5V. There should be a current between the two ( and therefore potential changes ), but not elsewhere. 

I tried setting a Neumann BC to 0, but it is not what I want to enforce. I don't want the net current exiting the simulation to be 0, I want the current exiting the simulation to be 0 everywhere. What is the fem / mfem function I should use for that ?

![image](https://user-images.githubusercontent.com/45109319/225614802-604553e2-8c3f-4019-af0d-8ef77e73b77c.png)

**Reply by jandrej** [reply]
If you have a "grounded outer layer", your boundary condition on the surface is Dirichlet = 0V. In 3D, depending on what your model should represent, then this means to also set the surface facing us and the one in the opposite to be zero. That would mean you can only see the potential if you slice into the visualized field.

**Reply by mlstowell** [reply]
To have zero current flow at a boundary the condition you want is $n\cdot\sigma\nabla\phi=0$. This is a homogeneous Neumann boundary condition and it happens to be the default boundary condition when using H1 finite elements with this operator so you don't need to do anything. I think @jandrej is on the right track, it looks like your outermost surfaces are being set to 0V using a Dirichlet Boundary condition. Remove these boundaries from your Dirichlet BC and it will probably work as expected.

**Reply by Drinausaur** [reply]
Oh, this was (again) a convergence problem. The code did converge, but the tolerance was not tight enough and lead to this strange result --'. Sorry
